hive-hooks/src/main/java/com/expediagroup/apiary/extensions/hooks/pathconversion/converters/PathConverter.java

```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+package com.expediagroup.apiary.extensions.hooks.pathconversion.converters;
+
+import java.util.regex.Matcher;
+
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+
+import lombok.extern.slf4j.Slf4j;
+
+import com.expediagroup.apiary.extensions.hooks.config.Configuration;
+import com.expediagroup.apiary.extensions.hooks.converters.GenericConverter;
+import com.expediagroup.apiary.extensions.hooks.pathconversion.config.PathConversionConfiguration;
+import com.expediagroup.apiary.extensions.hooks.pathconversion.models.PathConversion;
+
+@Slf4j
+public class PathConverter extends GenericConverter {
+
+  public PathConverter(Configuration configuration) {
+    super(configuration);
+  }
+
+  @Override
+  public PathConversionConfiguration getConfiguration() {
+    return (PathConversionConfiguration) super.getConfiguration();
+  }
+
+  /**
+   * Converts a path for the given Table.
+   *
+   * @param table Table location to potentially alter.
+   * @return true if table location is altered, false otherwise.
+   */
+  @Override
+  public boolean convertTable(Table table) {
+    if (!getConfiguration().isPathConversionEnabled()) {
+      log.trace("PathConversion is disabled. Skipping path conversion for table.");
+      return false;
+    }
+
+    StorageDescriptor sd = table.getSd();
+    log.debug("Examining table location: {}", sd.getLocation());
+    return convertStorageDescriptor(sd);
+  }
+
+  /**
+   * Converts a path for the given Partition.
+   *
+   * @param partition Partition location to potentially alter.
+   * @return true if partition location is altered, false otherwise.
+   */
+  @Override
+  public boolean convertPartition(Partition partition) {
+    if (!getConfiguration().isPathConversionEnabled()) {
+      log.trace("PathConversion is disabled. Skipping path conversion for partition.");
+      return false;
+    }
+
+    StorageDescriptor sd = partition.getSd();
+    log.debug("Examining partition location: {}", sd.getLocation());
+    return convertStorageDescriptor(sd);
+  }
+
+  @Override
+  public boolean convertStorageDescriptor(StorageDescriptor sd) {
+    boolean pathConverted = false;
+    for (PathConversion pathConversion : getConfiguration().getPathConversions()) {
+      Matcher matcher = pathConversion.pathPattern.matcher(sd.getLocation());
+      if (matcher.find()) {
+        for (Integer captureGroup : pathConversion.captureGroups) {
+          if (hasCaptureGroup(matcher, captureGroup, sd.getLocation())) {
+            String newLocation = sd.getLocation().replace(matcher.group(captureGroup), pathConversion.replacementValue);
+            log.info("Switching storage location {} to {}.", sd.getLocation(), newLocation);
+            sd.setLocation(newLocation);
+            pathConverted = true;
```

Comment:
If we were using this on arbitrary paths named like:
```
s3://mybucket-us-west-2/orig-location-us-west-2/table/data=<YYYY-MM-DD>
```
And wrote a regex to find all/both `us-west-2` occurrences, and replace them with `us-east-1` by specifying:
```
<property>
   <name>apiary.path.replacement.capturegroups.<MY_FOO_REPLACER></name>
   <value>1,2</value>
</property>
```
Then wouldn't the above code log multiple `Switching storage` messages like so? 
```
Switching storage location s3://mybucket-us-west-2/orig-location-us-west-2/table/data=<YYYY-MM-DD> to s3://mybucket-us-east1/orig-location-us-west-2/table/data=<YYYY-MM-DD>.
Switching storage location s3://mybucket-us-east1/orig-location-us-west-2/table/data=<YYYY-MM-DD> to s3://mybucket-us-east1/orig-location-us-east-1/table/data=<YYYY-MM-DD>.
```
That seems like the logs for that would be very confusing.

Reply:
Actually, looking at the code in line 87, it calls `String.replace(CharSequence, CharSequence)` which will replace all instances.  So if I changed `apiary.path.replacement.capturegroups.<MY_FOO_REPLACER>` to just `1`, this code looks like it will replace *both* instances of `us-west-2`, which would be incorrect.

Reply:
ah good catch. Will work on a fix for this

Reply:
Ok both of these cases should be accounted for now. Instead of a String.replace, if a match is found I convert the current sd location to a StringBuffer and use the match boundaries of the capture group to do the replace. I had to add a buffer offset here to account for multiple replacements that were larger / smaller